packages/mobx/__tests__/v5/base/autorunAsync.js

```diff
@@ -85,6 +85,53 @@ test("autorun 1", function (done) {
     }, 30)
 })
 
+test("autorun re-calculates computed states only when required", function (done) {
+    let autoruns = 0
+    let computedCalcs = 0
+
+    const o = mobx.observable.box(0)
+    const c = mobx.computed(() => {
+        computedCalcs++
+        return o.get()
+    })
+
+    mobx.autorun(
+        () => {
+            c.get()
+            autoruns++
+        },
+        { delay: 300 }
+    )
+    expect(computedCalcs).toBe(0)
+
+    setTimeout(() => {
+        // We expect the first run to run the computed
+        expect(autoruns).toBe(1)
+        expect(computedCalcs).toBe(1)
+    }, 400)
+
+    // We shouldn't eagerly calculate the state before the autorun runs because
+    // the state can be mutated again before `delay` expires - if that happens, a
+    // re-calculation here would be wasted.
+    setTimeout(() => {
+        o.set(1)
+        expect(computedCalcs).toBe(1)
+    }, 800)
+
+    // This will re-invalidate the state so we'll need to recalculate
+    // again when autorun runs (when `delay` expires).
+    setTimeout(() => {
+        o.set(3)
+        expect(computedCalcs).toBe(1)
+    }, 900)
+
+    setTimeout(() => {
+        expect(autoruns).toBe(2)
+        expect(computedCalcs).toBe(2)
+        done()
+    }, 1600)
+})
```

Comment:
Thanks @realyze - your [reproduction](https://codesandbox.io/p/sandbox/mobx-autorun-recalculation-issue-z3vggt?file=%2Fpackage.json%3A17%2C2) helped craft this test

Reply:
For reference to reviewers: the same test fails on the latest `main`:

![image](https://github.com/mobxjs/mobx/assets/13295571/1bdda0f5-d50b-46d7-a093-4652680bc358)